testing/base/source_gen.h

```diff
@@ -0,0 +1,227 @@
+//   - Imports and exports
+//   - API files and impl files.
+// - Be random but deterministic. The goal is benchmarking and so while this
+//   code should strive for not producing trivially predictable patterns, it
+//   should also strive to be consistent and suitable for benchmarking. Wherever
+//   possible, it should permute the order and content without randomizing the
+//   total count, size, or complexity.
+//
+// Note that the default and primary generation target is interesting Carbon
+// source code. We have a best-effort to alternatively generate comparable C++
+// constructs to the Carbon ones for comparative benchmarking, but there is no
+// goal to cover all the interesting C++ patterns we might want to benchmark,
+// and we don't aim for perfectly synthesizing C++ analogs. We can always drop
+// fidelity for the C++ code path if needed for simplicity.
+//
+// TODO: There are numerous places where we hard code a fixed quantity. Instead,
+// we should build a rich but general system to easily encode a discrete
+// distribution that is sampled. We have a specialized version of this for
+// identifiers that should be generalized.
+class SourceGen {
+ public:
+  enum class Language {
+    Carbon,
+    Cpp,
+  };
+
+  struct FunctionDeclParams {
+    // TODD: Arbitrary default, should switch to a distribution from data.
+    int max_params = 4;
+  };
+
+  struct MethodDeclParams {
+    // TODD: Arbitrary default, should switch to a distribution from data.
+    int max_params = 4;
+  };
+
+  // Parameters used to generate a class in a generated file.
+  //
+  // Currently, this uses a fixed number of each kind of declaration, with
+  // arbitrary defaults chosen. The defaults currently skew towards large
+  // classes with lots of nested declarations.
+  // TODO: Switch these to distributions based on data.
+  //
+  // TODO: Add support for generating definitions and parameters to control
+  // them.
+  struct ClassParams {
+    int public_function_decls = 4;
+    FunctionDeclParams public_function_decl_params = {.max_params = 8};
+
+    int public_method_decls = 10;
+    MethodDeclParams public_method_decl_params;
+
+    int private_function_decls = 2;
+    FunctionDeclParams private_function_decl_params = {.max_params = 6};
+
+    int private_method_decls = 8;
+    MethodDeclParams private_method_decl_params = {.max_params = 6};
+
+    int private_field_decls = 6;
+  };
+
+  // Parameters used to generate a file with dense declarations.
+  struct DenseDeclParams {
+    // TODO: Add more parameters to control generating top-level constructs
+    // other than class definitions.
+
+    // Parameters used when generating class definitions.
+    ClassParams class_params = {};
+  };
+
+  // Access a global instance of this type for use in benchmarks, tests, or
+  // other places where sharing a common instance is useful. Note that there is
+  // nothing thread safe about this instance or type.
+  static auto Global() -> SourceGen&;
```

Comment:
```suggestion
  // Access a global instance of this type to generate Carbon code for benchmarks, tests, or
  // other places where sharing a common instance is useful. Note that there is
  // nothing thread safe about this instance or type.
  static auto Global() -> SourceGen&;
```
Seems important to point out this global instance generates Carbon not C++.